sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
+        return self.connection.subprotocol
+
+    def pause_frames(self):
+        if not self.can_pause:
+            return False
+        if self.pause_frame_fut is not None:
+            return False
+        if self.loop is None or self.io_proto is None:
+            return False
+        if self.io_proto.transport is not None:
+            self.io_proto.transport.pause_reading()
+        self.pause_frame_fut = self.loop.create_future()
+        return True
+
+    def resume_frames(self):
+        if self.pause_frame_fut is None:
+            return False
+        if self.loop is None or self.io_proto is None:
+            error_logger.warning("Websocket attempting to resume reading frames, but connection is gone.")
+            return False
+        if self.io_proto.transport is not None:
+            self.io_proto.transport.resume_reading()
+        self.pause_frame_fut.set_result(None)
+        self.pause_frame_fut = None
+        return True
+
+    async def connection_made(self, io_proto: asyncio.Protocol, loop=None):
+        if loop is None:
+            try:
+                loop = getattr(io_proto, "loop")
+            except AttributeError:
+                loop = asyncio.get_event_loop()
+        self.loop = loop
+        self.io_proto = io_proto  # this will be a WebSocketProtocol
+        self.connection_lost_waiter = self.loop.create_future()
+        self.data_finished_fut = asyncio.shield(self.loop.create_future())
+
+        if self.ping_interval is not None:
+            self.keepalive_ping_task = asyncio.create_task(self.keepalive_ping())
+        self.close_connection_task = asyncio.create_task(self.auto_close_connection())
+
+    async def wait_for_connection_lost(self, timeout=10) -> bool:
+        """
+        Wait until the TCP connection is closed or ``timeout`` elapses.
+
+        Return ``True`` if the connection is closed and ``False`` otherwise.
+
+        """
+        if self.connection_lost_waiter is None:
+            return False
+        if not self.connection_lost_waiter.done():
+            try:
+                await asyncio.wait_for(
+                    asyncio.shield(self.connection_lost_waiter), timeout
+                )
+            except asyncio.TimeoutError:
+                pass
+        # Re-check self.connection_lost_waiter.done() synchronously because
+        # connection_lost() could run between the moment the timeout occurs
+        # and the moment this coroutine resumes running.
+        return self.connection_lost_waiter.done()
+
+    async def process_events(self, events: Sequence[Event]) -> None:
+        """
+        Process a list of incoming events.
+        """
+        # Wrapped in a mutex lock, to prevent other incoming events
+        # from processing at the same time
+        async with self.process_event_mutex:
+            for event in events:
+                if event.opcode == OP_PONG:
+                    await self.process_pong(event)
+                else:
```

Comment:
I would use `elif event.opcode in DATA_OPCODES` here, where `DATA_OPCODES` comes from `websockets.frames`. I should make it a public API.

Reply:
Is there a possibility of `event.opcode` _not_ being in `DATA_OPCODES`? 

If no, why the `elif`? Wouldn't we want everything that is not a `PONG` to fall thru to this condition?

If yes, then how do we handle that case?

Reply:
`DATA_OPCODES = OP_CONT, OP_TEXT, OP_BINARY`

=> you want to extract frames from these

`CTRL_OPCODES = OP_CLOSE, OP_PING, OP_PONG`

=> here you only care about OP_PONG because you provide an abstraction over ping / pong with a Future that completes when receiving a pong; websockets deals with OP_PING and OP_CLOSE

Reply:
Thanks. I was meaning to ask about that. That answers my question. I'll change it to `elif event.opcode in DATA_OPCODES`.